### DISCUSSION TITLE
JGroups JDBC_PING discovery protocol for Infinispan

### DISCUSSION THREAD

**Discussion by xgp** [keycloak/keycloak#10177] [Archive]
I've read the new documentation on setting up distributed caching (https://www.keycloak.org/server/caching). I'm unable to find information on support for using `JDBC_PING` discovery protocol for JGroups/Infinispan.

I used to set this up (via the Docker image) by setting the following two environment variables:
```
JGROUPS_DISCOVERY_PROTOCOL=JDBC_PING
JGROUPS_DISCOVERY_PROPERTIES=remove_all_data_on_view_change=true,clear_table_on_view_change=true,datasource_jndi_name=java:jboss/datasources/KeycloakDS,info_writer_sleep_time=500,initialize_sql=\"CREATE TABLE IF NOT EXISTS JGROUPSPING (own_addr varchar(200) NOT NULL, cluster_name varchar(200) NOT NULL, ping_data BYTEA, constraint PK_JGROUPSPING PRIMARY KEY (own_addr, cluster_name));\"
```

This doesn't appear to be supported any longer, as it was something that got processed in the "legacy" docker container. Also, it doesn't appear to have support as one of the "cache stacks" in the new Quarkus distribution.

Is there documentation on how to migrate from this discovery protocol, or is it possible to specify this in a custom `conf/cache-ispn.xml` file?

**Comment by xgp** [comment]
In working on this, I also encountered that there no longer appears to be a data source bound to JNDI for the Keycloak database. Is that correct?

**Comment by xgp** [comment]
I'm unable to get this addition to the conf/cache-ispn.xml file to load a new stack:
```xml
...
  <jgroups>
    <stack name="jdbc-ping" extends="tcp">
      <JDBC_PING connection_driver="org.postgresql.Driver"
                 connection_username="keycloak" connection_password="password"
                 connection_url="jdbc:postgres://postgres/keycloak" level="debug"
                 initialize_sql="CREATE TABLE IF NOT EXISTS JGROUPSPING (own_addr varchar(200) NOT NULL, cluster_name varchar(200) NOT NULL, ping_data BYTEA, constraint PK_JGROUPSPING PRIMARY KEY (own_addr, cluster_name));"
                 info_writer_sleep_time="500"
                 remove_all_data_on_view_change="true"
                 clear_table_on_view_change="true"
                 stack.combine="REPLACE"
                 stack.position="MPING" />
    </stack>
  </jgroups>
...
```
perhaps due to this issue? https://github.com/keycloak/keycloak/issues/10052

**Reply by thomasdarimont** [reply]
Try this with `"jdbc:postgresql://postgres/keycloak"` instead of `"jdbc:postgres://postgres/keycloak"`.
Also `clear_table_on_view_change="true"` and `level="debug"` seems to result in an error - try to remove those settings.

**Reply by xgp** [reply]
Thanks @thomasdarimont! I had that wrong, and was not specifying the custom protocol as the `stack` attribute of the `transport` element.

**Reply by pedroigor** [reply]
Thanks @thomasdarimont. And yes, you can define your own `stack` within the cache configuration file. You can do like you are doing (extending) or referencing a jgroups config file.

**Reply by datacoderid** [reply]
Hi @xgp , found some issue. Could you share how to custom cache-stack=name?

**Reply by thomasdarimont** [reply]
@datacoderid You can configure a custom stack configuration file with an env variable like `KC_CACHE_CONFIG_FILE=cache-ispn-jdbc-ping.xml`.
Take a look at the following gist for an example: https://gist.github.com/thomasdarimont/47fc89ca8f51f5142418c3318c0d2e82

**Reply by pedroigor** [reply]
@thomasdarimont Would make sense to make this file available by default in the distribution?

**Reply by thomasdarimont** [reply]
@pedroigor yes I think this would be useful, I'll send a PR for this. 

Additionaly support for KUBE_PING would also be helpful, but this requires an additional jar (jgroups-kubernetes-1.0.16.jar) in the /providers (or lib/lib/main) folder. I have a working configuration for this too. Would that also be an option?

**Reply by datacoderid** [reply]
Hi @thomasdarimont , 

I still got this error, even though I already set `--cache-config-file` and `--cache=ispn`
I believe I missed something, do you have any clue?
Btw, I am using standard installation (binary file) not Docker

(keycloak-cache-init) ISPN000569: Unable to persist Infinispan internal caches as no global state enabled
[org.jgroups.protocols.pbcast.GMS] (keycloak-cache-init) uat-uat-vm-sso-vm-1-14186: JOIN(uat-uat-vm-sso-vm-9161d2-1-14186) sent to uat-uat-vm-sso-vm-9161d2-0-52466 timed out (after 2000 ms), on try 9

**Reply by datacoderid** [reply]
nvm, the issue solved after I request to open all firewall ports between both VMs.
thanks all.. Now my issue in LB, but will discuss it outside this discussion

**Reply by pedroigor** [reply]
> @pedroigor yes I think this would be useful, I'll send a PR for this.
> 
> Additionaly support for KUBE_PING would also be helpful, but this requires an additional jar (jgroups-kubernetes-1.0.16.jar) in the /providers (or lib/lib/main) folder. I have a working configuration for this too. Would that also be an option?

@thomasdarimont I think we should be fine for k8s, no? The stack `kubernetes` is based on ISPN defaults.

**Reply by thomasdarimont** [reply]
The [default kubernetes stack uses DNS_PING](https://github.com/infinispan/infinispan/blob/main/core/src/main/resources/default-configs/default-jgroups-kubernetes.xml#L27), which works well in practice but is often replaced with kube_ping in customer projects.

A working version for using KUBE_PING with Keycloak can be found here: https://github.com/codecentric/helm-charts/blob/master/charts/keycloakx/examples/postgresql-kubeping/cache-ispn-kubeping.xml#L12

See: http://jgroups.org/manual4/index.html#_kube_ping

**Reply by pedroigor** [reply]
AFAIK, DNS_PING is the preferable discovery protocol for k8s. At least that is the feedback we have from ISPN team.

Could you elaborate why DNS_PING is not enough? Trying to understand if we need to review our defaults.

**Comment by stianst** [comment]
I'm not sure if JDBC_PING is really something we should support at all, as it's mostly a workaround. What environment are you running in? Are there other protocols you could consider instead?

**Reply by xgp** [reply]
There are 30+ installations that I have personally done using JDBC_PING.  Some examples are:
- AWS EC2/ECS/Fargate where embedding of AWS credentials (even for locked-down users) is not allowed by policy.
- On-prem native workloads where the network does not allow multicast.
- On-prem container workloads that are not k8s (Nomad, Mesos, Swarm, etc.) where the network does not allow multicast.

Also, any installations where the goal is the *reduce* the number of external dependencies (a completely reasonable customer request), I have found that most gravitate to using JDBC_PING. My experience operating it has also been that it is reliable for moderately sized clusters, and standardization on it has made my and my customer administrator's lives easier.

From the volume of messages on the various lists regarding people using JDBC_PING, I'm guessing I'm not the only one "doing it wrong". If Keycloak doesn't want to support it, that's obviously up to you. But if you do decide to support custom Infinispan configurations, could we keep this discussion alive for the purpose of sharing information on a common user need? Thank you.

**Reply by dasniko** [reply]
JDB_PING is one of the most flexible protocols, it's easy to use and works in nearly every environment. Therefore, I vote for keeping and supporting it!

I can nearly second @xgp's arguments, and I also do have a couple of customers using JDBC_PING very successfully!

Think of a on-premise environment, w/o OpenShift/K8s, not other Cloud-provider available/usable, no Multicast allowed/working, no dedicated hostnames/ip-addresses b/c of DNS, but the hosts use different hostnames/IPs, so querying the DNS does not make sense... What should customers do?

**Reply by laskasn** [reply]
We also use on almost all of our production installations the JDBC_PING protocol for the node discovery. I completely agree with @xgp and @dasniko (they explain perfectly why multicast in practice actually proves insufficient), therefore, you should double think before deciding to drop the jdbc_ping support. 
I believe that if we had a poll about "which method do you use for infinispan nodes' discovery?"  you would be surprised of how many are actually using the JDBC_PING!

**Reply by kkcmadhu** [reply]
Whats the recommended stuff for k8s deployments @stianst  dns ping, tcp ping??

**Comment by xgp** [comment]
Full working example here https://gist.github.com/xgp/768eea11f92806b9c83f95902f7f8f80

**Reply by xgp** [reply]
Note that this example currently requires your postgres credentials to be placed in your custom cache configuration file, which is a bad practice. Unless there is some way to specify environment variables in these config files that I don't know about, I'd still like to see better support for things like this that doesn't require the embedding of credentials, as this is also the reason several customers don't use things like S3_PING.

**Reply by thomasdarimont** [reply]
You can use env variables in custom jgroups stack files like this:
```
    <JDBC_PING connection_driver="org.postgresql.Driver"
               connection_username="${KC_DB_USERNAME}"
               connection_password="${KC_DB_PASSWORD}"
               connection_url="jdbc:postgresql://${KC_DB_URL_HOST}/${KC_DB_DATABASE}"
               initialize_sql="CREATE TABLE IF NOT EXISTS JGROUPSPING (own_addr varchar(200) NOT NULL, cluster_name varchar(200) NOT NULL, ping_data BYTEA, constraint PK_JGROUPSPING PRIMARY KEY (own_addr, cluster_name));"
               info_writer_sleep_time="500"
               remove_all_data_on_view_change="true"
    />
```

Infinispan complains, but jgroups resolves the variable references correctly.
```
2022-02-14 22:36:55,278 WARN  [org.infinispan.commons.util.StringPropertyReplacer] (keycloak-cache-init) ISPN000901: Property KC_DB_PASSWORD could not be replaced as intended!
2022-02-14 22:36:55,279 WARN  [org.infinispan.commons.util.StringPropertyReplacer] (keycloak-cache-init) ISPN000901: Property KC_DB_URL_HOST could not be replaced as intended!
2022-02-14 22:36:55,279 WARN  [org.infinispan.commons.util.StringPropertyReplacer] (keycloak-cache-init) ISPN000901: Property KC_DB_DATABASE could not be replaced as intended!
2022-02-14 22:36:55,279 WARN  [org.infinispan.commons.util.StringPropertyReplacer] (keycloak-cache-init) ISPN000901: Property KC_DB_USERNAME could not be replaced as intended!
```

but jgroups resolves those variables.

**Reply by thomasdarimont** [reply]
You could try to prefix the env variable lookup with `env.`, e.g.: `${env.KC_DB_USERNAME}`

**Reply by xgp** [reply]
Thanks. The `env.` prefix is working. For some circumstances it worked without it, but in others it threw an exception.

**Reply by 3XC1T3D** [reply]
@xgp do you habe some experience after two month of running keycloak with your jdbc_ping setup?

I have created a new discussion if there are some best practise for this scenario -> https://github.com/keycloak/keycloak/discussions/11272

**Reply by initshdb** [reply]
@thomasdarimont @xgp is there a way to make the database configuration to be read in the `cache-ispn.xml` from the `conf/keycloak.conf` file instead of the environment variables - `${env.KC_DB_USERNAME}`. I not running Keycloak in docker, and have all my configuration in the `conf/keycloak.conf` file.

**Reply by xgp** [reply]
@nitishborade I'm not aware of a way to have it come from the `keycloak.conf` file. 

Even if you are not running in Docker, you can still set the appropriate environment variables so that they are converted by `${env.xyz..}`. In cases where I am not using Docker, I usually just write a startup script that sets environment variables before calling `kc.sh ...`.

**Reply by initshdb** [reply]
@xgp Thanks for the prompt help. Yes setting as environment variables in the a startup script helps, however i just wanted to check if whether there was an approach to read from the `keycloak.conf`. 
Nevertheless, thank you for the this alternative.

**Comment by Arulaln-AR** [comment]
HI @xgp @pedroigor @thomasdarimont,

Do you guys have an example for aws s3 ping, we have an keycloak installed on top EKS cluster and worker nodes are based on ec2 instances.

I followed the article "https://gist.github.com/thomasdarimont/47fc89ca8f51f5142418c3318c0d2e82" and "https://gist.github.com/xgp/768eea11f92806b9c83f95902f7f8f80". Can set the jdbc_ping stack and table and records are created in the DB.

I am trying to setup the aws s3 ping for my actual environment, I followed an article in keycloak but struck without the complete steps.
It would be great if you have guys have done the similar setup at your end and need the steps for the same.

**Comment by kkcmadhu** [comment]
For k8s Dns ping/k8s ping can be natural choice.. irrespecitve.of aks/eks.
not sure why you eant to use s3 ping... (s3 ping too should work though)..


Sent from Yahoo Mail on Android 
 
  On Tue, 26 Jul 2022 at 4:58 pm, ***@***.***> wrote:   


HI @xgp @pedroigor @thomasdarimont,

Do you guys have an example for aws s3 ping, we have an keycloak installed on top EKS cluster and worker nodes are based on ec2 instances.

I followed the article "https://gist.github.com/thomasdarimont/47fc89ca8f51f5142418c3318c0d2e82" and "https://gist.github.com/xgp/768eea11f92806b9c83f95902f7f8f80". Can set the jdbc_ping stack and table and records are created in the DB.

I am trying to setup the aws s3 ping for my actual environment, I followed an article in keycloak but struck without the complete steps.
It would be great if you have guys have done the similar setup at your end and need the steps for the same.

**Comment by Arulaln-AR** [comment]
Hi Madhu,

I followed the article "https://www.keycloak.org/server/caching" where
under section "*Additional transport stacks*" they mentioned that for
Amazon EC2 instances i need to use "ec2" stack. We have an EKS cluster
where worker nodes are built using Amazon EC2 instances. Hence I thought we
should use the "ec2" stack. Correct me if I'm wrong please.



On Wed, 27 Jul 2022 at 10:00, kkcmadhu ***@***.***> wrote:

> For k8s Dns ping/k8s ping can be natural choice.. irrespecitve.of aks/eks.
> not sure why you eant to use s3 ping... (s3 ping too should work though)..
>
>
> Sent from Yahoo Mail on Android
>
> On Tue, 26 Jul 2022 at 4:58 pm, ***@***.***> wrote:
>
>
> HI @xgp @pedroigor @thomasdarimont,
>
> Do you guys have an example for aws s3 ping, we have an keycloak installed
> on top EKS cluster and worker nodes are based on ec2 instances.
>
> I followed the article "
> https://gist.github.com/thomasdarimont/47fc89ca8f51f5142418c3318c0d2e82"
> and "https://gist.github.com/xgp/768eea11f92806b9c83f95902f7f8f80". Can
> set the jdbc_ping stack and table and records are created in the DB.
>
> I am trying to setup the aws s3 ping for my actual environment, I followed
> an article in keycloak but struck without the complete steps.
> It would be great if you have guys have done the similar setup at your end
> and need the steps for the same.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.Message ID: ***@***.***>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keycloak/keycloak/discussions/10177#discussioncomment-3257269>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AP4YZ5J7E3GWXP23ONCW7Q3VWC3NVANCNFSM5OHQJRUA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Regards,
Arulaln A R
+91-8754438576

**Comment by Arulaln-AR** [comment]
Hi Madhu,

Could you please check the below query. When I run the keycloak using EKS
cluster, can I still use the kubernetes stack?

On Wed, 27 Jul 2022 at 13:34, Arulaln A R ***@***.***> wrote:

> Hi Madhu,
>
> I followed the article "https://www.keycloak.org/server/caching" where
> under section "*Additional transport stacks*" they mentioned that for
> Amazon EC2 instances i need to use "ec2" stack. We have an EKS cluster
> where worker nodes are built using Amazon EC2 instances. Hence I thought we
> should use the "ec2" stack. Correct me if I'm wrong please.
>
>
>
> On Wed, 27 Jul 2022 at 10:00, kkcmadhu ***@***.***> wrote:
>
>> For k8s Dns ping/k8s ping can be natural choice.. irrespecitve.of aks/eks.
>> not sure why you eant to use s3 ping... (s3 ping too should work though)..
>>
>>
>> Sent from Yahoo Mail on Android
>>
>> On Tue, 26 Jul 2022 at 4:58 pm, ***@***.***> wrote:
>>
>>
>> HI @xgp @pedroigor @thomasdarimont,
>>
>> Do you guys have an example for aws s3 ping, we have an keycloak
>> installed on top EKS cluster and worker nodes are based on ec2 instances.
>>
>> I followed the article "
>> https://gist.github.com/thomasdarimont/47fc89ca8f51f5142418c3318c0d2e82"
>> and "https://gist.github.com/xgp/768eea11f92806b9c83f95902f7f8f80". Can
>> set the jdbc_ping stack and table and records are created in the DB.
>>
>> I am trying to setup the aws s3 ping for my actual environment, I
>> followed an article in keycloak but struck without the complete steps.
>> It would be great if you have guys have done the similar setup at your
>> end and need the steps for the same.
>>
>> —
>> Reply to this email directly, view it on GitHub, or unsubscribe.
>> You are receiving this because you commented.Message ID: ***@***.***>
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/keycloak/keycloak/discussions/10177#discussioncomment-3257269>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AP4YZ5J7E3GWXP23ONCW7Q3VWC3NVANCNFSM5OHQJRUA>
>> .
>> You are receiving this because you commented.Message ID:
>> ***@***.***>
>>
>
>
> --
> Regards,
> Arulaln A R
> +91-8754438576
>


-- 
Regards,
Arulaln A R
+91-8754438576

**Comment by kkcmadhu** [comment]
EKS is kubenetes(managed k8s) , so dns ping, k8s ping should just work oob for you.
Refer this thread for details Configuring distributed caches · Discussion #12022 · keycloak/keycloak

| 
| 
| 
|  |  |

 |

 |
| 
|  | 
Configuring distributed caches · Discussion #12022 · keycloak/keycloak

Hello, I managed to do this configuration in the Wildfly distribution. I can't do it in the Quarkus distribution...
 |

 |

 |




Madhu
    On Thursday, 28 July, 2022 at 08:08:47 pm IST, Arulaln-AR ***@***.***> wrote:  
 
 

Hi Madhu,

Could you please check the below query. When I run the keycloak using EKS
cluster, can I still use the kubernetes stack?

On Wed, 27 Jul 2022 at 13:34, Arulaln A R ***@***.***> wrote:

> Hi Madhu,
>
> I followed the article "https://www.keycloak.org/server/caching" where
> under section "*Additional transport stacks*" they mentioned that for
> Amazon EC2 instances i need to use "ec2" stack. We have an EKS cluster
> where worker nodes are built using Amazon EC2 instances. Hence I thought we
> should use the "ec2" stack. Correct me if I'm wrong please.
>
>
>
> On Wed, 27 Jul 2022 at 10:00, kkcmadhu ***@***.***> wrote:
>
>> For k8s Dns ping/k8s ping can be natural choice.. irrespecitve.of aks/eks.
>> not sure why you eant to use s3 ping... (s3 ping too should work though)..
>>
>>
>> Sent from Yahoo Mail on Android
>>
>> On Tue, 26 Jul 2022 at 4:58 pm, ***@***.***> wrote:
>>
>>
>> HI @xgp @pedroigor @thomasdarimont,
>>
>> Do you guys have an example for aws s3 ping, we have an keycloak
>> installed on top EKS cluster and worker nodes are based on ec2 instances.
>>
>> I followed the article "
>> https://gist.github.com/thomasdarimont/47fc89ca8f51f5142418c3318c0d2e82"
>> and "https://gist.github.com/xgp/768eea11f92806b9c83f95902f7f8f80". Can
>> set the jdbc_ping stack and table and records are created in the DB.
>>
>> I am trying to setup the aws s3 ping for my actual environment, I
>> followed an article in keycloak but struck without the complete steps.
>> It would be great if you have guys have done the similar setup at your
>> end and need the steps for the same.
>>
>> —
>> Reply to this email directly, view it on GitHub, or unsubscribe.
>> You are receiving this because you commented.Message ID: ***@***.***>
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/keycloak/keycloak/discussions/10177#discussioncomment-3257269>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AP4YZ5J7E3GWXP23ONCW7Q3VWC3NVANCNFSM5OHQJRUA>
>> .
>> You are receiving this because you commented.Message ID:
>> ***@***.***>
>>
>
>
> --
> Regards,
> Arulaln A R
> +91-8754438576
>


-- 
Regards,
Arulaln A R
+91-8754438576

**Comment by kkcmadhu** [comment]
S3 ping are right choice if you are running keycloak on bare bone ec2.. since you are on 3ks, dns ping or k8s ping might work well ..

Sent from Yahoo Mail on Android 
 
  On Wed, 27 Jul 2022 at 1:35 pm, ***@***.***> wrote:   

Hi Madhu,

I followed the article "https://www.keycloak.org/server/caching" where
under section "*Additional transport stacks*" they mentioned that for
Amazon EC2 instances i need to use "ec2" stack. We have an EKS cluster
where worker nodes are built using Amazon EC2 instances. Hence I thought we
should use the "ec2" stack. Correct me if I'm wrong please.



On Wed, 27 Jul 2022 at 10:00, kkcmadhu ***@***.***> wrote:

> For k8s Dns ping/k8s ping can be natural choice.. irrespecitve.of aks/eks.
> not sure why you eant to use s3 ping... (s3 ping too should work though)..
>
>
> Sent from Yahoo Mail on Android
>
> On Tue, 26 Jul 2022 at 4:58 pm, ***@***.***> wrote:
>
>
> HI @xgp @pedroigor @thomasdarimont,
>
> Do you guys have an example for aws s3 ping, we have an keycloak installed
> on top EKS cluster and worker nodes are based on ec2 instances.
>
> I followed the article "
> https://gist.github.com/thomasdarimont/47fc89ca8f51f5142418c3318c0d2e82"
> and "https://gist.github.com/xgp/768eea11f92806b9c83f95902f7f8f80". Can
> set the jdbc_ping stack and table and records are created in the DB.
>
> I am trying to setup the aws s3 ping for my actual environment, I followed
> an article in keycloak but struck without the complete steps.
> It would be great if you have guys have done the similar setup at your end
> and need the steps for the same.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.Message ID: ***@***.***>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keycloak/keycloak/discussions/10177#discussioncomment-3257269>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AP4YZ5J7E3GWXP23ONCW7Q3VWC3NVANCNFSM5OHQJRUA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Regards,
Arulaln A R
+91-8754438576
